main.py

```diff
@@ -77,33 +97,57 @@ def export_streams_to_clipboard(self):
 
     def add_new_stream(self, *args, stream_url=None, stream_quality=cfg[CONFIG_QUALITY]):
         """Adds a new player for the specified stream in the grid."""
+
         if not stream_url:
             stream_url, ok = QtWidgets.QInputDialog.getText(self, "Stream input", "Enter the stream URL:")
 
             if not ok:
                 return
 
+        # Give some feedback to the user
+        self.grid.addWidget(self.loading, self.coordinates.x, self.coordinates.y)
+        self.loading.show()
+
+        # Run the rest on a seperate thread to be able to show the loading feedback
+        # Also helps a lot with lag
+        threading.Thread(target=self._add_new_stream, args=(stream_url, stream_quality)).start()
+
+    def _add_new_stream(self, stream_url=None, stream_quality=cfg[CONFIG_QUALITY]):
+        """Fetches qualities and if possible adds a frame to the main window."""
         try:
             stream_options = self.streamlink_session.streams(stream_url)
 
             if stream_quality not in stream_options:
-                stream_quality, ok = self._get_user_quality_preference(stream_options)
-
-                if not ok:
-                    return
+                self.fail_add_stream.emit("default_quality_missing", (stream_options, stream_url, stream_quality))
+                return
 
-            frame = self.setup_videoframe(stream_options, stream_quality, self.coordinates)
-            self.update_new_stream_coordinates()
-            frame.player.audio_set_mute(cfg[CONFIG_MUTE])
+            self.add_frame.emit(stream_options, stream_quality, self.coordinates)
 
         except streamlink.exceptions.NoPluginError:
-            error_window = QtWidgets.QMessageBox().warning(
-                self,
-                "Error",
-                "Could not open stream: The provided URL is not supported"
+            self.fail_add_stream.emit(
+                "url_not_supported",
+                (
+                    "Error",
+                    "Could not open stream: The provided URL is not supported"
+                )
             )
 
-            self.add_new_stream()
+    def on_fail_add_stream(self, err, args):
+        # Remove the loading feedback
+        self.grid.removeWidget(self.loading)
+        self.loading.hide()
+
+        if err == "url_not_supported":
```

Comment:
I was unsure of the python way of doing this. Since I did not want to clutter the code with lots of signals, I decided this was the better way to go.